lib/sneakers/handlers/maxretry.rb

```diff
@@ -165,9 +165,14 @@ def failure_count(headers)
         if headers.nil? || headers['x-death'].nil?
           0
         else
-          headers['x-death'].select do |x_death|
+          x_death_array = headers['x-death'].select do |x_death|
             x_death['queue'] == @worker_queue_name
-          end.count
+          end
+          if x_death_array.count != 1
+            x_death_array.count
+          else
+            x_death_array.first['count'] || 1
```

Comment:
Your interpretation for the "retired once case" is correct.  It might be a little more clear if the logic was refactored to something like this:

```ruby
if x_death_array.count == 1 && x_death_array.first['count']
  x_death_array.first['count']
else
  x_death_array.count
end
```

For your second question, the x-death headers are always an array.  If there isn't an entry, then it will return x_death_array.count which will be 0, which is the same thing that happens in the current version of the code.  This method is returning the failure count, which will always be 0 the first time it's in the queue.